persist/json.go

```diff
@@ -1,93 +1,238 @@
 package persist
 
 import (
+	"bytes"
 	"encoding/json"
-	"io"
+	"io/ioutil"
 	"os"
+	"strings"
+
+	"github.com/NebulousLabs/Sia/build"
+	"github.com/NebulousLabs/Sia/crypto"
 )
 
-// Load loads json data from a reader.
-func Load(meta Metadata, data interface{}, r io.Reader) error {
+// readJSON will try to read a persisted json object from a file.
+func readJSON(meta Metadata, object interface{}, filename string) error {
+	// Open the file.
+	file, err := os.Open(filename)
+	if os.IsNotExist(err) {
+		return err
+	}
+	if err != nil {
+		return build.ExtendErr("unable to open persisted json object file", err)
+	}
+	defer file.Close()
+
+	// Read the metadata from the file.
 	var header, version string
-	dec := json.NewDecoder(r)
+	dec := json.NewDecoder(file)
 	if err := dec.Decode(&header); err != nil {
-		return err
+		return build.ExtendErr("unable to read header from persisted json object file", err)
 	}
 	if header != meta.Header {
 		return ErrBadHeader
 	}
 	if err := dec.Decode(&version); err != nil {
-		return err
+		return build.ExtendErr("unable to read version from persisted json object file", err)
 	}
 	if version != meta.Version {
 		return ErrBadVersion
 	}
-	if err := dec.Decode(data); err != nil {
-		return err
+
+	// Read everything else.
+	remainingBytes, err := ioutil.ReadAll(dec.Buffered())
+	if err != nil {
+		return build.ExtendErr("unable to read persisted json object data", err)
 	}
+	// The buffer may or may not have read the rest of the file, read the rest
+	// of the file to be certain.
+	remainingBytesExtra, err := ioutil.ReadAll(file)
+	if err != nil {
+		return build.ExtendErr("unable to read persisted json object data", err)
+	}
+	remainingBytes = append(remainingBytes, remainingBytesExtra...)
 
-	return nil
+	// Determine whether the leading bytes contain a checksum. A proper checksum
+	// will be 67 bytes (quote, 64 byte checksum, quote, newline). A manual
+	// checksum will be the characters "manual\n" (9 characters). If neither
+	// decode correctly, it is assumed that there is no checksum at all.
+	var checksum crypto.Hash
+	err = json.Unmarshal(remainingBytes[:67], &checksum)
+	if err == nil && checksum == crypto.HashBytes(remainingBytes[68:]) {
+		// Checksum is proper, and matches the data. Update the data portion to
+		// exclude the checksum.
+		remainingBytes = remainingBytes[68:]
+	} else {
+		// Cryptographic checksum failed, try interpreting a manual checksum.
+		var manualChecksum string
+		err := json.Unmarshal(remainingBytes[:8], &manualChecksum)
+		if err == nil && manualChecksum == "manual" {
+			// Manual checksum is proper. Update the remaining data to exclude
+			// the manual checksum.
+			remainingBytes = remainingBytes[9:]
+		}
+	}
```

Comment:
doesn't seem like the "checksum exists, but is wrong" case is handled explicitly